File: MemoizationSearch/Memoization.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.Serialization.Formatters.Binary;

public class Memoizer<TResult> where TResult : new()  // This constraint ensures TResult has a parameterless constructor
{
    private readonly Func<object[], TResult> func;
    private readonly Dictionary<Tuple<object[]>, (TResult Result, ulong Timestamp)> cache;
    private ulong timeout;

    public Memoizer(Func<object[], TResult> function, ulong timeout)
    {
        this.func = function;
        this.cache = new Dictionary<Tuple<object[]>, (TResult Result, ulong Timestamp)>(new TupleComparer());
        this.timeout = timeout;
    }

    public TResult Call(params object[] args)
    {
        ulong currentMillis = (ulong)DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var key = Tuple.Create(args);

        if (cache.ContainsKey(key))
        {
            (TResult result, ulong timestamp) = cache[key];
            if (currentMillis - timestamp < timeout || timeout == ulong.MaxValue)
                return result;
        }

        TResult newResult = func(args);
        cache[key] = (newResult, currentMillis);
        return newResult;
    }
    public TResult Seek(params object[] args)
    {
        // Create a key from the arguments array
        var key = Tuple.Create(args);

        // Check if the key exists in the cache
        if (cache.ContainsKey(key))
        {
            // Retrieve the cached result
            (TResult result, ulong timestamp) = cache[key];
            // Optionally, update timestamp or add logic to handle stale data
            return result;
        }

        // Return a new instance of TResult if not found in cache
        return new TResult();
    }
    private class TupleComparer : IEqualityComparer<Tuple<object[]>>
    {
        public bool Equals(Tuple<object[]> x, Tuple<object[]> y)
        {
            return x.Item1.SequenceEqual(y.Item1);
        }

        public int GetHashCode(Tuple<object[]> obj)
        {
            unchecked // Overflow is fine, just wrap
            {
                int hash = 19;
                foreach (var item in obj.Item1)
                {
                    hash = hash * 31 + (item?.GetHashCode() ?? 0);
                }
                return hash;
            }
        }
    }
    public void ClearCache(params object[] args)
    {
        var key = Tuple.Create(args);
        if (cache.ContainsKey(key))
        {
            cache.Remove(key);
        }
    }


    public void SaveToFile(string filename)
    {
        using (var stream = File.OpenWrite(filename))
        {
            var formatter = new BinaryFormatter();
            formatter.Serialize(stream, cache);
        }
    }

    public void LoadFromFile(string filename)
    {

            using (var stream = File.OpenRead(filename))
            {
                var formatter = new BinaryFormatter();
                var deserializedData = (Dictionary<Tuple<object[]>, (TResult Result, ulong Timestamp)>)formatter.Deserialize(stream);
                cache.Clear();
                foreach (var item in deserializedData)
                {
                    cache.Add(item.Key, item.Value);
                }
            }
    }


    public void SetCacheTimeout(ulong newTimeout)
    {
        timeout = newTimeout;
    }

    public ulong GetCacheTimeout()
    {
        return timeout;
    }

    public int CacheSize()
    {
        return cache.Count;
    }

    public override string ToString()
    {
        return $"Cache contains {CacheSize()} items.";
    }
}

File: MemoizationSearch/memoization.py
import pickle
import time
import random
cachetime=200
INFINITYCACHE=0xffffffff
class Memoizer:
    def __init__(self, func, timeout=cachetime):
        self.func = func
        self.cache = {}
        self.timeout = timeout
    def __call__(self, *args):
        current_time = time.time_ns() // 1_000_000 #ms
        if args in self.cache:
            result, timestamp = self.cache[args]
            if current_time - timestamp < self.timeout:
                return result
        result = self.func(*args)
        self.cache[args] = (result, current_time)
        return result
    def ClearCache(self):
        self.cache.clear()
    def ClearCache(self,*args):
        del self.cache[args]
    def getrandomNumbers(self):
        return random.randint(0, 100)
    def save_to_file(self, filename):
        with open(filename, 'wb') as f:
            pickle.dump(self.cache, f)
        print(f"Cache saved to {filename}")
    def load_from_file(self, filename):
        try:
            with open(filename, 'rb') as f:
                self.cache = pickle.load(f)
            print(f"Cache loaded from {filename}")
        except FileNotFoundError:
            print(f"No cache file found at {filename}")
    def __rshift__(self, file):
         pickle.dump(self.cache, file)
    def __lshift__(self, data):
        self.cache = pickle.load(data)
    def seek(self,*args):
        if args in self.cache:
            result, timestamp = self.cache[ args]
            return result
        return None;
    def SetCacheResetTime(self,retvalue,validtime,*args):
        self.cache[args] = {retvalue,validtime}
    def SetCacheTime(self,time):
        self.timeout=time;
    def GetCacheTime(self):
        return self.timeout
    def __iter__(self):
        return iter(self.cache.items())
    def __str__(self):
        return f"Cache contains {self.size()} items."
    def empty(self):
        return self.size()
    def size(self):
        return len(self.cache)

